Mlem/App/Views/Shared/Search/SearchSheetView.swift

```diff
@@ -0,0 +1,88 @@
+//
+//  SearchSheetView.swift
+//  Mlem
+//
+//  Created by Sjmarf on 27/06/2024.
+//
+
+import Combine
+import MlemMiddleware
+import SwiftUI
+
+struct SearchSheetView<Item: Searchable, Content: View>: View {
+    @Environment(AppState.self) var appState
+    @Environment(\.dismiss) var dismiss
+    
+    enum CloseButtonLabel: String {
+        case cancel, done
+    }
+    
+    @ViewBuilder let content: ([Item], DismissAction) -> Content
+    let closeButtonLabel: CloseButtonLabel
+    
+    @State var query: String = ""
+    @State var results: [Item] = []
+    
+    @State var editing: Bool = true
+    @State var focused: Bool = true
+    
+    init(
+        closeButtonLabel: CloseButtonLabel = .cancel,
+        @ViewBuilder content: @escaping ([Item], DismissAction) -> Content
+    ) {
+        self.content = content
+        self.closeButtonLabel = closeButtonLabel
+    }
+    
+    var body: some View {
+        ScrollView {
+            VStack(alignment: .leading, spacing: 0) {
+                content(results, dismiss)
+            }
+        }
+        .navigationBarTitleDisplayMode(.inline)
+        .toolbar {
+            ToolbarItem(placement: .principal) {
+                HStack(spacing: 0) {
+                    SearchBar("Search", text: $query, isEditing: $editing)
+                        .isInitialFirstResponder(true)
+                        .focused($focused)
+                }
+            }
+            ToolbarItem(placement: .topBarTrailing) {
+                Button(closeButtonLabel.rawValue.capitalized) {
+                    dismiss()
+                }
+            }
+        }
+        .task(id: query, priority: .userInitiated) { @MainActor in
+            do {
+                if !query.isEmpty {
+                    try await Task.sleep(for: .seconds(0.2))
+                }
+                results = try await Item.search(
+                    api: appState.firstApi,
+                    query: query,
+                    page: 1,
+                    limit: 20
+                )
+            } catch {
+                handleError(error)
+            }
```

Comment:
This is a ton of work to put on the `MainActor`--everything except the actual assignments to `results` should happen on a background thread (especially `Task.sleep`). You can feel the hang if you enter a query and immediately dismiss the sheet--it's reasonably benign here because there aren't other animations to interrupt, but still a pattern to avoid.

```suggestion
        .task(id: query, priority: .userInitiated) {
            do {
                if !query.isEmpty {
                    try await Task.sleep(for: .seconds(0.2))
                }
                let response = try await Item.search(
                    api: appState.firstApi,
                    query: query,
                    page: 1,
                    limit: 20
                )
                Task { @MainActor in
                    results = response
                }
            } catch {
                handleError(error)
            }
        }
```